File: engineering/source/scripts/verify-checksums.py
from pathlib import Path
import hashlib, sys
ROOT=Path(__file__).resolve().parents[1]
manifest=ROOT/'CHECKSUMS_SHA256.txt'
errors=[]; checked=0
for line_no,line in enumerate(manifest.read_text(encoding='utf-8').splitlines(),1):
    if not line.strip(): continue
    try: expected, rel=line.split('  ',1)
    except ValueError:
        errors.append(f'line {line_no}: malformed'); continue
    path=ROOT/rel
    if not path.is_file():
        errors.append(f'missing: {rel}'); continue
    h=hashlib.sha256()
    with path.open('rb') as f:
        for chunk in iter(lambda:f.read(1024*1024),b''): h.update(chunk)
    actual=h.hexdigest(); checked+=1
    if actual!=expected: errors.append(f'mismatch: {rel}: {actual} != {expected}')
if errors:
    print('\n'.join(errors)); print(f'CHECKSUM FAIL: {checked} checked, {len(errors)} errors'); sys.exit(1)
print(f'CHECKSUM PASS: {checked} payload files')

File: engineering/source/scripts/verify-checksums.mjs
import { createHash } from 'node:crypto';
import { readFile, readdir, stat } from 'node:fs/promises';
import { dirname, join, relative, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const manifestPath = join(root, 'CHECKSUMS_SHA256.txt');
const manifest = await readFile(manifestPath, 'utf8');
const expected = new Map();
for (const line of manifest.split(/\r?\n/)) {
  if (!line.trim()) continue;
  const match = line.match(/^([0-9a-f]{64})  (.+)$/);
  if (!match) throw new Error(`Malformed checksum line: ${line}`);
  if (expected.has(match[2])) throw new Error(`Duplicate path: ${match[2]}`);
  expected.set(match[2], match[1]);
}

async function walk(dir) {
  const result = [];
  for (const name of await readdir(dir)) {
    if (name === 'node_modules' || name === '.npm-cache' || name.startsWith('.browser-home') || name === '__pycache__') continue;
    const path = join(dir, name);
    const info = await stat(path);
    if (info.isDirectory()) result.push(...await walk(path));
    else if (info.isFile()) result.push(path);
  }
  return result;
}
const files = (await walk(root)).filter(path => relative(root, path).replaceAll('\\', '/') !== 'CHECKSUMS_SHA256.txt');
const actualPaths = new Set(files.map(path => relative(root, path).replaceAll('\\', '/')));
for (const path of expected.keys()) if (!actualPaths.has(path)) throw new Error(`Missing file: ${path}`);
for (const path of actualPaths) if (!expected.has(path)) throw new Error(`Unlisted file: ${path}`);
for (const path of [...expected.keys()].sort()) {
  const digest = createHash('sha256').update(await readFile(join(root, path))).digest('hex');
  if (digest !== expected.get(path)) throw new Error(`Checksum mismatch: ${path}`);
}
console.log(`CHECKSUM VERIFICATION PASS: ${expected.size} files`);
